File: src/processOrderVanillaSync.ts
import { orders, books } from './data'
import { validateOrder, Order, SyncProcessor, PlacedOrderResult } from './api'

const bookService = (bookId: string) => (books[bookId] ? books[bookId] : null)

const orderService = (orderId: string) =>
  orders[orderId] ? orders[orderId] : null

const validationService = (order: Order) => {
  return validateOrder(order)
}

const calculateAmountService = (order: Order) => {
  let total = 0
  for (let i = 0; i < order.items.length; i++) {
    const item = order.items[i]
    const book = bookService(item.bookId)
    if (book != null) {
      total += item.quantity * book.price
    } else {
      throw new Error('Book not found: ' + item.bookId)
    }
  }
  return total
}

const placeOrderService = (order: Order) => {
  let totalAmount = calculateAmountService(order)
  return {
    success: true,
    totalAmount
  }
}

const processor: SyncProcessor = (orderId: string): PlacedOrderResult => {
  const order = orderService(orderId)
  if (order == null) {
    return {
      success: false,
      totalAmount: 0.0
    }
  }

  const validationResult = validationService(order)
  if (!validationResult.valid) {
    return {
      success: false,
      totalAmount: 0.0
    }
  }

  return placeOrderService(order)
}

export default processor

File: demo/rust/runner/src/process_order_vanilla_sync.rs
use crate::api::*;
use crate::data::{get_book, get_order};

fn book_service(id: &String) -> Option<&'static Book> {
    get_book(id)
}

fn order_service(id: &String) -> Option<&'static Order> {
    get_order(id)
}

fn validation_service(order: &Order) -> Result<&Order, OrderNotValid> {
    match validate_order(order) {
        Ok(_) => Ok(order),
        Err(err) => Err(err),
    }
}

fn calculate_amount_service(order: &Order) -> f64 {
    let mut total = 0.0;
    for item in &order.items {
        let book = book_service(&item.book_id);
        match book {
            Some(b) => {
                total += item.quantity as f64 * b.price;
            }
            _ => {}
        };
    }
    total
}

fn place_order_service(order: &Order) -> PlacedOrderResult {
    let amount = calculate_amount_service(order);
    Ok(OrderSuccessful::new(amount))
}

pub struct VanillaProcessorSync {}

impl SyncProcessor for VanillaProcessorSync {
    fn process(&self, order_id: &String) -> Result<f64, ()> {
        match order_service(order_id) {
            Some(order) => {
                let validation = validation_service(&order);
                match validation {
                    Ok(order) => match place_order_service(order) {
                        Ok(res) => Ok(res.amount),
                        Err(_) => Err(()),
                    },
                    _ => Err(()),
                }
            }
            _ => Err(()),
        }
    }
}

impl VanillaProcessorSync {
    pub fn processor() -> &'static dyn SyncProcessor {
        &(VanillaProcessorSync {}) as &dyn SyncProcessor
    }
}

pub fn process_syncv_direct(order_id: &String) -> Result<f64, ()> {
    match order_service(order_id) {
        Some(order) => {
            let validation = validation_service(&order);
            match validation {
                Ok(order) => match place_order_service(order) {
                    Ok(res) => Ok(res.amount),
                    Err(_) => Err(()),
                },
                _ => Err(()),
            }
        }
        _ => Err(()),
    }
}
